### ISSUE TITLE
Support Assistants API

### ISSUE THREAD

**Issue by benedictstrube** [langchain4j/langchain4j#271]
Support the new Assistants API Beta of Open AI.

**Comment by remyohajinwa** [comment]
Hey is anyone working on this. I would like to take a look @benedictstrube ?

**Comment by clun** [comment]
I noticed the openai integration leverages openai4j library
https://github.com/ai-for-java/openai4j

I do not have anything against this but it is only 22 stars and unfortunately lagging behind. Why not moving to the [official one](https://github.com/TheoKanning/openai-java) i could help porting it TBH. Also the later provides the Assistant API.

**Comment by langchain4j** [comment]
@clun https://github.com/ai-for-java/openai4j is our own OpenAI client and we can change it as required. For now we will stick to it due to multiple reason, one of them is that it allows to integrate langchain4j into Quarkus ecosystem nicely. It is not a matter of a missing functionality in openai4j (this can be added easily), but more about how this Assistants API fits into LangChain4j.

**Comment by langchain4j** [comment]
Hi @benedictstrube, could you please provide a bit more details on the motivation?
As far as I see it, OpenAI takes the abstractions that we already have (chat history, agent, tools, RAG, etc) and hides it all behind their API. Similar level of abstraction can be achieved with `AiServices` in LangChain4j right now. There seems to be nothing that LangChain4j does not support already. But on the downside, it removes all the control from you (you don't have an option to customize anything) and costs quite a lot.
Supporting Assistants API will require quite a lot of work from our side and it would be awesome to first understand the needs better. Thanks.

**Comment by langchain4j** [comment]
The general sentiment is that OpenAI's Assistants API is an expensive black box that you cannot customize (at least right now):
https://www.reddit.com/r/LangChain/comments/185cdot/pros_and_cons_of_relying_on_the_new_openai/
https://www.reddit.com/r/LangChain/comments/18e25qx/openai_assistant_pricing_vs_api_pricing/
https://www.reddit.com/r/LangChain/comments/17pbynv/new_openai_assistant_api/
https://www.reddit.com/r/OpenAI/comments/18xk1ap/assistant_api_with_gpt4_is_expensive/
https://www.reddit.com/r/OpenAI/comments/17ykj1h/how_is_assistants_going_for_you/

**Comment by langchain4j** [comment]
More:
https://www.reddit.com/r/ChatGPTCoding/comments/1akctxu/my_mixed_thoughts_on_openais_assistants_api/
https://www.reddit.com/r/ChatGPTCoding/comments/19841qr/openai_assistants_api_yes_or_no/
https://www.reddit.com/r/OpenAI/comments/18xr4kv/is_the_assistants_api_a_trap/

**Comment by langchain4j** [comment]
Great news, everyone!
@new-koncept will contribute and oversee support for the Assistants API in LangChain4j.
Stay tuned, it will be available soon!

**Comment by fgreco55** [comment]
> Great news, everyone! @new-koncept will contribute and oversee support for the Assistants API in LangChain4j. Stay tuned, it will be available soon!

Sounds great.  But since the Assistants API is relegated to OpenAI (afaik), is building an abstraction truly worth the effort?

**Comment by langchain4j** [comment]
@fgreco55 I am not sure it will be an abstraction, perhaps just a client. But this API is actually supported also by Azure OpenAI and I've seens some open-source implementations (e.g. from DataStax)

**Comment by fgreco55** [comment]
Thanks for the info.

But if the OpenAI Assistants API basically gives you RAG (or maybe RAG-ish) behavior on your retrieval docs that you upload to Altman, why would I use the usual LC4J API for context memory, completion, context window control, et al... other than portability to other LLM services?

Btw, the Assistants run-steps logging looks quite useful for auditing and regulatory mandates.

**Comment by langchain4j** [comment]
> docs that you upload to Altman

😆 

> But if the OpenAI Assistants API basically gives you RAG (or maybe RAG-ish) behavior on your retrieval docs that you upload to Altman, why would I use the usual LC4J API for context memory, completion, context window control, et al... other than portability to other LLM services?

Good question. Perhaps @new-koncept can answer it better. From my pow, the idea was to provide a Java client so that Assistants API can be used from Java apps. But that was before OpenAI introduced an official Java/Kotlin SDK, so perhaps it makes less sense right now.

**Comment by fgreco55** [comment]
That's exactly what I mean.  OpenAI has an early official Java/Kotlin API, so presumably, developing another API might be a waste of effort.  Duplicating the Assistants API so it would work across LLMs would be cool, but that implies a back-end service that provides data persistence, data security, code interpreter and sandbox, et al... a non-trivial exercise.

Btw, would you trust Altman with your data?  I wouldn't.

**Comment by alfonsomunozpomer** [comment]
I was wondering what the status is regarding this issue. Has it been decided to skip this because of OpenAI’s official library?
Thanks for the effort!

**Comment by dliubarskyi** [comment]
@alfonsomunozpomer yes, please use the official OpenAI java sdk

**Comment by MikhailShumenko** [comment]
This repository was archived by the owner on Feb 11, 2025. It is now read-only.

https://github.com/ai-for-java/openai4j

Should we use this one: https://github.com/openai/openai-java ?

BTW, are you going to migrate langchain4j core classes to using the official open-ai java library?

**Comment by dliubarskyi** [comment]
@MikhailShumenko yes, please use https://github.com/openai/openai-java for Assistants API

> BTW, are you going to migrate langchain4j core classes to using the official open-ai java library?

Yes, there is a [PR](https://github.com/langchain4j/langchain4j/pull/2508) to support new official OpenAI SDK. Existing OpenAI module (`langchain4j-open-ai`) will keep using our own OpenAI integration which enables better integration into Quarkus and Spring projects
